### PR TITLE
Add lightsail_static_ip module

### LINKED ISSUE #250: Add a module to manage AWS Lightsail Static IPs
##### SUMMARY
Issue #174 requests adding the ability to manage firewall and static IP settings in a Lightsail VPS. I suggest that it may be beneficial to have a module that can manage static IPs for Lightsail (which are managed seperately from VMs).

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
community.aws.lightsail

##### ADDITIONAL INFORMATION
In Lightsail, static IP addresses are managed separately from VPSes. You create and manage static IPs, and can attach them to a VPS as per #174. A module that can mange these IPs would be a prerequisite for end-to-end configuration management for Lightsail.

I already have a basic version of this and am planning on submitting a PR in the next couple of weeks once it's cleaned up & tested

Relevant API Documentation:

* https://docs.aws.amazon.com/lightsail/2016-11-28/api-reference/API_AllocateStaticIp.html
* https://docs.aws.amazon.com/lightsail/2016-11-28/api-reference/API_GetStaticIps.html
* https://docs.aws.amazon.com/lightsail/2016-11-28/api-reference/API_ReleaseStaticIp.html

Possible Playbook notation:
```yaml
- name: Create a new Lightsail static IP
  community.aws.lightsail_static_ip:
    state: present
    name: static_ip_1
    aws_access_key: XXXXXXXX
```

### PR DESCRIPTION
**Pull request by danielcotton**
##### SUMMARY
Adds a module to manage AWS Lightsail Static IP addresses

Fixes #250 

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
community.aws.lightsail_static_ip

##### ADDITIONAL INFORMATION
As per my earlier pull request, this is part of larger work to add functionality to this module in terms of managing AWS Lightsail.

Apologies if there are any issues/I have missed anything. I am new to Ansible development, but have tried to follow the contribution docs as best as I can.

Example playbook:
```paste below
    - name: create static IP
      community.aws.lightsail_static_ip:
        name: "test_static"
        state: present
        region: ap-southeast-2
```

### CODE REVIEW

**Comment by danielcotton** [comment]
It would seem integration tests are failing due to insufficient permissions on the test runner's IAM user. I presume this isn't something I can fix myself. Could a maintainer please investigate?

The new module uses the following actions:
lightsail:GetStaticIp
lightsail:AllocateStaticIp
lightsail:ReleaseStaticIp

Example error message from Shippable:
`
AccessDeniedException: An error occurred (AccessDeniedException) when calling the GetStaticIp operation: User: arn:aws:sts::966509639900:assumed-role/ansible-core-ci-test-prod/prod=shippable=ansible-collections=community.aws=862.22 is not authorized to perform: lightsail:GetStaticIp on resource: arn:aws:lightsail:us-east-1:966509639900:*
`

I will fix sanity test issues as soon as I get time in the next day or so

**Comment by danielcotton** [comment]
Hi @jillr, I believe the IAM user for integration tests may need to have additional permissions added for tests to pass.
This new module uses:
* lightsail:GetStaticIp
* lightsail:AllocateStaticIp 
* lightsail:ReleaseStaticIp

**Comment by tremble** [comment]
Hi @danielcotton 

We generally recommend that the contributor raise a PR to get the permissions changed: https://docs.ansible.com/ansible/latest/dev_guide/platforms/aws_guidelines.html#aws-permissions-for-integration-tests

You can find the existing lightsail permissions in the 'paas' policy:
https://github.com/mattclay/aws-terminator/blob/master/aws/policy/paas.yaml#L42

**Comment by danielcotton** [comment]
Oh, thank you for that @tremble, I'd missed that doc and assumed permission changes were only available to maintainers.

I will raise a PR over there. Thanks again :)

**Comment by tremble** [comment]
I've just noticed that we don't currently link to it from our README in this collection (and that it's still missing a few collections migration related changes), I'll raise a PR to update that.

File: tests/integration/targets/lightsail_static_ip/defaults/main.yml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+static_ip_name: "{{ resource_prefix }}_static_ip"
+region: "us-west-2"
```
Comment (jillr) [inline tests/integration/targets/lightsail_static_ip/defaults/main.yml:2]:
```suggestion
```
Our CI uses us-east-1, and tests should generally run in whatever region is in the configuration supplied by the user running them (usually via cloud-config-aws.ini).

File: tests/integration/targets/lightsail_static_ip/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+      region: '{{ aws_region | default(omit) }}'
+
+  block:
+
+    # ==== Tests ===================================================
+
+    - name: Create a new static IP
+      lightsail_static_ip:
+        name: "{{ static_ip_name }}"
+        region: "{{ region }}"
```
Comment (jillr) [inline tests/integration/targets/lightsail_static_ip/tasks/main.yml:17]:
```suggestion
```
Region is set my module_defaults, you don't need to define it for any tasks.

File: tests/integration/targets/lightsail_static_ip/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+      register: result
+
+    - assert:
+        that:
+          - result.changed == True
+          - "'static_ip' in result and result.static_ip.name == static_ip_name"
+          - "'static_ip' in result and result.static_ip.resourceType == 'StaticIP'"
```
Comment (jillr) [inline tests/integration/targets/lightsail_static_ip/tasks/main.yml:24]: This assertion fails, the key and value being returned are `"resource_type": "StaticIp",`

File: tests/integration/targets/lightsail_static_ip/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+          - "result.instance.state.name == 'running'"
```
Comment (jillr) [inline tests/integration/targets/lightsail_static_ip/tasks/main.yml:25]: `allocate_static_ip_address` does not return a state key.  It does return a `status` but the module does not directly return this response. The keys being returned by `get_static_ip` that can be asserted look like this:
```
    "static_ip": {
        "arn": "arn:aws:lightsail:us-east-1:966509639900:StaticIp/9658cdc0-fef3-4fdd-83f3-2a438a8608a4",
        "created_at": "2021-01-14T21:38:48.416000+00:00",
        "ip_address": "54.236.109.178",
        "is_attached": false,
        "location": {
            "availability_zone": "all",
            "region_name": "us-east-1"
        },
        "name": "ansible-test-lab-36684871_static_ip",
        "resource_type": "StaticIp",
        "support_code": "011036789062/54.236.109.178"
    }
```

**Review by jillr — DISMISSED** [review]

**Comment by danielcotton** [comment]
Hi @jillr @tremble, it looks like CI cleanup failed so you may have some Lightsail static IPs left dangling on the account. It is probably worth checking and cleaning these up as they become billable after a certain period.

Is there any way to run integration tests locally (preferably for a specific module)?

**Comment by jillr** [comment]
Thanks @danielcotton, there's steps for how to test locally in this post (near the bottom the header is "Testing Locally")  https://www.ansible.com/blog/getting-started-with-aws-ansible-module-development

**Comment by danielcotton** [comment]
Hi @jillr. Thanks for that article - it cleared up a lot of confusion I had about setting up a local dev environment.

I believe I've resolved the issues. It looks like Shippable tests [26](https://app.shippable.com/github/ansible-collections/community.aws/runs/1682/26/tests) and [27](https://app.shippable.com/github/ansible-collections/community.aws/runs/1682/27/tests) failed saying a region is required, but this doesn't occur on other tests, so seems isolated to 2.9 but I'm not sure why it is failing

**Comment by tremble** [comment]
2.9 tests are failing because it's not been added to the 2.9 defaults file (2.9 doesn't read from runtime.yml, the 2.9 defaults are hard-coded into ansible)

@jillr we discussed skipping 2.9 for new modules...

**Comment by danielcotton** [comment]
So is there anything further I can do to help get this merged? Or are we just waiting on changes to the test matrix?

**Review by tremble — APPROVED** [review]

**Comment by tremble** [comment]
@danielcotton I'm sorry it's taken so long to get this merged.  We fixed the integration tests a little while back and I forgot to come back and review this.

Many thanks for your time on this PR.

**Comment by danielcotton** [comment]
All good @tremble, thank you and @jillr for your help with my first Ansible PR

**Pull request merged**